### DISCUSSION TITLE
When next release?

### DISCUSSION THREAD

**Discussion by ghost** [shouldly/shouldly#837] [General]
There has been a fair amount of fixes since the last release (more than 1 year ago), wouldn't it be time to release a new minor version?
For example, this [one](https://github.com/shouldly/shouldly/pull/790) is a much needed improvement.

**Comment by jnm2** [comment]
That's a good point. We followed up with https://github.com/shouldly/shouldly/releases/tag/4.1.0 in August.

**Comment by jarlef** [comment]
Why are pull requests being merged but there are no more releases going out? The last nuget package is from early 2023. Thinking about switching from FluentAssertion but a bit wary to switch to a library that are slow on updating. Is it due to lack of maintainers?

**Reply by slang25** [reply]
It's more that we are figuring out where to draw the line for releases. I had some things in mind for a 5.0 release, but given the sudden rising adoption of the library we should probably get a 4.3.0 out and then re-evaluate.

**Comment by SimonCropp** [comment]
i am happy with a more regular cadence of releases. for all my other projects i usually do a deploy every week if there have been any features or bug fixes in that week

**Comment by jarlef** [comment]
My main concerns in just the ability to get important patches and upgrades. E.g upgrading libraries with vulnerabilities, security fixes and also targeting new frameworks e.g .NET 9, Xunit etc. Sounds like you could maybe benefit from having two branches with new features (dev) and the stable branch (main).

**Comment by jnm2** [comment]
What if we make the decisionmaking process easier and set expectations by setting a release cadence? By default there will be a release, but it can be skipped if it turns out there were no enhancements _and_ no vulnerable dependencies to update.

**Reply by SimonCropp** [reply]
sounds good to me

**Reply by slang25** [reply]
I'm thinking we release changes when we think it makes sense, but if nothing has been release for 3 months (maybe 6?), we force a release.

Personally I'd be against something as frequent as weekly, I think about how many projects use Shouldly, how many dependabot builds fire, how many automated downstream releases that triggers that then trigger more dependabot PRs... etc... Just at my company, the compute wastage would be significant.

**Reply by jnm2** [reply]
3 months matches the frequency of .NET SDKs and VS. JetBrains and DevExpress are generally twice a year, sometimes three times a year.

Can we try 3 months? Given today's release, that would mean we would not consider releases until the end of April, unless we judge something to be urgent enough to be a special case. Then ideally at the end of April we would check in and see if releasing would bring meaningful impact. We could weigh Shouldly changes and dependency vulnerabilities, for instance. This gives us some structure without waiting for agreement:

- If someone asks for a release sooner than end of April, and we aren't sure about this, we can say no, not before our cadence unless more motivation comes in.
- If someone asks for a release sooner than end of April and one of us really wants to do one, we can still do one.
- It would be good to have a calendar reminder to do that check-in every three months and automatically perform the release unless you suspect there's nothing worthy of release. (As in, maybe we refactored something, but there's literally nothing that affects end users.) Does anyone volunteer to drive this?
  - If we leave this up to being prompted by repo activity, we will miss the fact that users will be waiting for a release without explicitly asking for one, because they will be see our shipping cadence.
- If we decide to skip a release, or we forget to check in, and someone asks for a release after that, we could go ahead and do the "skipped" release at that point rather than waiting for the next three months to come around.

The goal with this would not be to restrain anyone's judgment calls, but just to reach a point where we are regularly releasing new features without having to make a multi-person decision, since we're busy folks. And to give folks an idea of what to expect.

**Reply by slang25** [reply]
Agreed, I've added a GHA workflow that will create an issue if there's been no release for 3 months:
https://github.com/shouldly/shouldly/blob/master/.github/workflows/release-reminder.yml

**Reply by jnm2** [reply]
That's such a good idea!
